File: diet-tracker/python/routers/diet.py
from fastapi import APIRouter, Depends, Request, Form, Query
from fastapi.responses import HTMLResponse, RedirectResponse
from fastapi.templating import Jinja2Templates
from sqlalchemy.orm import Session
from datetime import date as date_class
from models import SessionLocal, Meal, DailyLog, UserProfile
from services.diet_service import get_daily_summary
from routers.auth import get_current_user

router = APIRouter()
templates = Jinja2Templates(directory="templates")


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.get("/diet", response_class=HTMLResponse)
def show_diet(request: Request, date: str = None, db: Session = Depends(get_db), user: str = Depends(get_current_user)):
    target_date = date or str(date_class.today())
    profile = db.query(UserProfile).filter(UserProfile.user_id == user).first() or UserProfile(
        user_id=user, weight=70.0, height=175.0, age=25, gender="M"
    )
    
    meals = db.query(Meal).all()
    
    logs = db.query(DailyLog).filter(DailyLog.user_id == user, DailyLog.date == target_date).all()
    meal_ids = list(map(lambda l: l.meal_id, logs))
    logged_meals = db.query(Meal).filter(Meal.id.in_(meal_ids)).all() if meal_ids else []

    summary = get_daily_summary(user, target_date)

    return templates.TemplateResponse("diet.html", {
        "request": request,
        "user": user,
        "profile": profile,
        "meals": meals,
        "loggedMeals": logged_meals,
        "date": target_date,
        "summary": summary,
        "message": None
    })


@router.post("/diet/meal/add")
def add_meal_ui(date: str = Form(...), meal_id: int = Form(...), db: Session = Depends(get_db), user: str = Depends(get_current_user)):
    db.add(DailyLog(date=date, meal_id=meal_id, user_id=user))
    db.commit()
    return RedirectResponse(url=f"/diet?date={date}", status_code=303)


@router.post("/diet/profile/update")
def update_profile_ui(
    weight: float = Form(...),
    height: float = Form(...),
    age: int = Form(...),
    gender: str = Form(...),
    date: str = Form(...),
    db: Session = Depends(get_db),
    user: str = Depends(get_current_user)
):
    profile = db.query(UserProfile).filter(UserProfile.user_id == user).first()
    
    # Wykorzystujemy zaktualizowanie lub wstawienie nowego profilu
    db.merge(UserProfile(user_id=user, weight=weight, height=height, age=age, gender=gender))
    db.commit()
    
    return RedirectResponse(url=f"/diet?date={date}", status_code=303)


@router.get("/api/diet/logs/summary")
def api_get_summary(
    date: str = Query(None, description="Data w formacie YYYY-MM-DD"),
    user: str = Depends(get_current_user)
):
    """
    Pobiera podsumowanie makroskładników oraz realizację zapotrzebowania BMR dla danego dnia.
    """
    target_date = date or str(date_class.today())
    return get_daily_summary(user, target_date)

File: diet-tracker/nodejs/routes/diet.js
const express = require('express');
const router = express.Router();
const { Meal, DailyLog, UserProfile } = require('../models');
const { getDailySummary } = require('../services/dietService');
const { auth } = require('../middleware/auth');

// Widok główny aplikacji dietetycznej
router.get('/diet', auth, async (req, res) => {
  const today = new Date().toISOString().split('T')[0];
  const date = req.query.date || today;

  const profile = await UserProfile.findByPk(req.auth.user) || { weight: 70, height: 175, age: 25, gender: 'M' };
  const meals = await Meal.findAll();
  
  // Pobieramy posiłki przypisane do danego dnia
  const logs = await DailyLog.findAll({ where: { userId: req.auth.user, date } });
  const loggedMeals = await Meal.findAll({ where: { id: logs.map(l => l.mealId) } });

  const summary = await getDailySummary(req.auth.user, date);

  res.render('diet', {
    user: req.auth.user,
    profile,
    meals,
    loggedMeals,
    date,
    summary,
    message: null
  });
});

// Dodanie posiłku do dziennika
router.post('/diet/meal/add', auth, async (req, res) => {
  const { date, mealId } = req.body;
  await DailyLog.create({ date, mealId: parseInt(mealId), userId: req.auth.user });
  res.redirect(`/diet?date=${date}`);
});

// Aktualizacja parametrów wagi/wzrostu/wieku/płci
router.post('/diet/profile/update', auth, async (req, res) => {
  const { weight, height, age, gender, date } = req.body;
  
  await UserProfile.upsert({
    userId: req.auth.user,
    weight: parseFloat(weight),
    height: parseFloat(height),
    age: parseInt(age),
    gender
  });

  res.redirect(`/diet?date=${date}`);
});

/**
 * @openapi
 * /api/diet/logs/summary:
 *   get:
 *     summary: Pobierz podsumowanie makroskładników
 *     description: Pobiera podsumowanie dziennego zapotrzebowania (BMR), spożytych kalorii i makroskładników dla zalogowanego użytkownika w wybranym dniu.
 *     security:
 *       - basicAuth: []
 *     parameters:
 *       - in: query
 *         name: date
 *         required: true
 *         schema:
 *           type: string
 *         description: Data w formacie YYYY-MM-DD
 *     responses:
 *       200:
 *         description: Pomyślnie pobrano podsumowanie.
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 bmr:
 *                   type: number
 *                 proteins:
 *                   type: number
 *                 carbs:
 *                   type: number
 *                 fats:
 *                   type: number
 *                 calories:
 *                   type: number
 *                 percentage:
 *                   type: number
 */
router.get('/api/diet/logs/summary', auth, async (req, res) => {
  const date = req.query.date || new Date().toISOString().split('T')[0];
  const summary = await getDailySummary(req.auth.user, date);
  res.json(summary);
});

module.exports = router;
